release-common-lib/src/main/java/org/reactome/util/compare/DBObjectComparer.java

```diff
@@ -147,225 +176,428 @@ public static int compareInstances(
+	 * @param maxRecursionDepth The maximum depth of recursion that will be allowed. Normally a depth of 2 or 3 is
+	 * probably sufficient.
+	 * @param customAttributeNameFilter A custom Predicate that will be used to filter attribute names. The default
+	 * predicate looks like this:<pre>
+	Predicate&lt;? super SchemaAttribute&gt; attributeNameFilter = a -&gt; {
+	return !a.getName().equals("DB_ID")
+	&& !a.getName().equals("dateTime")
+	&& !a.getName().equals("modified")
+	&& !a.getName().equals("created");
+	};
+	</pre>
+	 * @param checkReferrers <code>true</code> if referrer attributes should be included; <code>false</code> otherwise
+	 * @return The total number of differences between the values compared (after recursion for GKInstance values)
+	 */
+	private static int compareValuesOfAttributeBetweenInstances(
+		SchemaAttribute attribute, Object value1, Object value2, GKInstance instance1, GKInstance instance2,
+		StringBuilder stringBuilder, int diffCount, int recursionDepth, int maxRecursionDepth,
+		Predicate<? super SchemaAttribute> customAttributeNameFilter, boolean checkReferrers
+	) {
+		int count = diffCount;
+		// Deal with attributes that return GKInstance objects.
+		if (isAttributeContainingInstances(attribute))
+		{
+			if (recursionDepth < maxRecursionDepth)
+			{
+				stringBuilder.append(getIndentString(recursionDepth))
+					.append(" Recursing on ")
+					.append(attribute.getName())
+					.append(" attribute...")
+					.append(System.lineSeparator());
+
+				return compareInstances(
+					(GKInstance) value1, (GKInstance) value2, stringBuilder, count, recursionDepth + 1,
+					maxRecursionDepth, customAttributeNameFilter, checkReferrers
+				);
+			}
+		}
+		// Deal with attributes that return "simple" things (Strings, numbers, etc..., arrays of
+		// Strings/numbers/etc...)
+		else if (!value1.equals(value2))
+		{
+			stringBuilder.append(
+				getIndentString(recursionDepth) + "Mismatch on" +
+					getAttributeRelationshipType(instance1, attribute) + " attribute " +
+					"'" + attribute.getName() + "'" + System.lineSeparator() +
+					getIndentString(recursionDepth) + "Instance 1 ('" + instance1 + "') has value:\t" +
+					value1 + System.lineSeparator() +
+					getIndentString(recursionDepth) + "Instance 2 ('" + instance2 + "') has value:\t" +
+					value2 + System.lineSeparator()
+			);
+			count++;
+		}
+		return count;
+	}
+
+	/**
+	 * Returns <code>true</code> if the attribute passed is an "instance attribute" (i.e. its value(s) is/are
+	 * instance(s)); <code>false otherwise</code>
+	 * @param attribute Attribute to check for the type of value it holds
+	 * @return <code>true</code> if the attribute passed holds instances; <code>false</code> otherwise
+	 */
+	private static boolean isAttributeContainingInstances(SchemaAttribute attribute)
+	{
+		Class<?> attributeType = attribute.getType();
+
+		return attributeType.equals(GKInstance.class) || attributeType.equals(Instance.class);
```

Comment:
Is there an actual difference between type `GKInstance` and type `Instance`? How are they employed differently

Reply:
It looks like `Instance` is the interface type and `GKInstance` is the class that implements the interface.  I'm not sure if there is another class that implements `Instance`.

@SolomonShorser-OICR I'm not too sure how `SomeType.class` works.  Is it possible for a Class object to hold the value of an interface and `attribute.getType()` would be equal to `Instance.class`?

Reply:
I can't recall how `.class` works in the context of interfaces. It should be possible to write a very small piece of test code to see if `GKInstance.class == Instance.class` (or is `.equals()` more appropriate here?), or just print GKInstance.class and Instancee.class to the console and see what you get. 

It might also depend on how the underlying Reactome library populates `attribute`...

Reply:
It looks like they are different returning the following for `GKInstance.class.toString()` and`Instance.class.toString()` respectively:

- class org.gk.model.GKInstance
- interface org.gk.model.Instance

I'm not sure what calls the setter corresponding to `getType()` method for `attribute` to fill its type, but I think the code should okay as is.  Checking both the `GKInstance` class and `Instance` interface may not be necessary but covers more bases in case `Instance.class` is used.